tests/unit/cpu_profiler_test.cc

```diff
@@ -0,0 +1,174 @@
+/*
+ * This file is open source software, licensed to you under the terms
+ * of the Apache License, Version 2.0 (the "License").  See the NOTICE file
+ * distributed with this work for additional information regarding copyright
+ * ownership.  You may not use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+/*
+ * Copyright (C) 2023 ScyllaDB Ltd.
+ */
+
+#include <boost/test/tools/old/interface.hpp>
+#include <cstddef>
+#include <seastar/core/internal/cpu_profiler.hh>
+#include <seastar/core/internal/stall_detector.hh>
+#include <seastar/core/reactor.hh>
+#include <seastar/core/thread_cputime_clock.hh>
+#include <seastar/core/loop.hh>
+#include <seastar/util/later.hh>
+#include <seastar/testing/test_case.hh>
+#include <seastar/testing/thread_test_case.hh>
+#include <atomic>
+#include <chrono>
+#include <sys/mman.h>
+
+#include "stall_detector_test_utilities.hh"
+
+struct temporary_profiler_settings {
+    std::chrono::nanoseconds prev_ns;
+    bool prev_enabled;
+
+    temporary_profiler_settings(bool enable, std::chrono::nanoseconds ns) {
+        prev_ns = engine().get_cpu_profiler_period();
+        prev_enabled = engine().get_cpu_profiler_enabled();
+
+        engine().set_cpu_profiler_period(ns);
+        engine().set_cpu_profiler_enabled(enable);
+    }
+
+    ~temporary_profiler_settings() {
+        engine().set_cpu_profiler_period(prev_ns);
+        engine().set_cpu_profiler_enabled(prev_enabled);
+    }
+};
+
+// If we set a timer to fire in N ms we can expect it to fire between N and (N + E) ms from
+// when it was set. Where E is some fixed error that arises from how granular a given timer
+// is. Therefore over a given period of time, M, if the timer is continously reset every
```

Comment:
It's not just the granularity of the timer, it's just an inherent problem with setting the next fire point based on a relative "P ns from now" API, isn't it?

Even if the timer was exact, the handler takes non-zero time from now so there will always be drift from the idea schedule (which I think is probably fine).

Reply:
Indeed, `E` could technically include the `P` for the time it takes to process the interrupt and set the timer again.